finality-aleph/src/data_io.rs

```diff
@@ -1,13 +1,190 @@
+    ) -> Self {
+        DataStore {
+            next_message_id: 0,
+            client,
+            message_requirements: HashMap::new(),
+            dependent_messages: HashMap::new(),
+            pending_messages: HashMap::new(),
+            available_blocks: LruCache::new(AVAILABLE_BLOCKS_CACHE_SIZE),
+            ready_messages_tx,
+            messages_rx,
+            _phantom: PhantomData,
+        }
+    }
+
+    pub(crate) async fn run(&mut self, mut exit: oneshot::Receiver<()>) {
+        let mut timeout = Delay::new(TIMEOUT_DURATION);
+        let mut import_stream = self.client.import_notification_stream();
+        loop {
+            tokio::select! {
+                Some(message) = &mut self.messages_rx.next() => {
+                    trace!(target: "afa", "Received message at Data Store {:?}", message);
+                    self.add_message(message);
+                }
+                Some(block) = &mut import_stream.next() => {
+                    trace!(target: "afa", "Block import notification at Data Store for block {:?}", block);
+                    self.add_block(block.hash);
+                }
+                _ = &mut timeout => {
+                    trace!(target: "afa", "Data Store timeout");
+                    let keys : Vec<_> = self.dependent_messages.keys().cloned().collect();
+                    for block_hash in keys {
+                        if let Ok(Some(_)) = self.client.header(BlockId::Hash(block_hash)) {
+                            self.add_block(block_hash);
+                        }
+                    }
+                    timeout = Delay::new(TIMEOUT_DURATION);
+                }
+                _ = &mut exit => {
+                    break;
+                }
+            }
+        }
+    }
+
+    fn forget_message(&mut self, message_id: MessageId) {
+        self.message_requirements.remove(&message_id);
+        if let Some(message) = self.pending_messages.remove(&message_id) {
+            for block_hash in message.included_data() {
+                if let Entry::Occupied(mut entry) = self.dependent_messages.entry(block_hash) {
+                    entry.get_mut().remove(&message_id);
+                    if entry.get().is_empty() {
+                        entry.remove_entry();
+                    }
+                }
+            }
+        }
+    }
+
+    fn add_pending_message(&mut self, message: AlephNetworkData<B>, requirements: Vec<B::Hash>) {
+        let message_id = self.next_message_id;
+        self.next_message_id += 1;
+        for block_hash in requirements.iter() {
+            self.dependent_messages
+                .entry(*block_hash)
+                .or_insert_with(HashSet::new)
+                .insert(message_id);
+        }
+        self.message_requirements
+            .insert(message_id, requirements.len());
+        self.pending_messages.insert(message_id, message);
+
+        if message_id >= PENDING_MESSAGE_LIFETIME {
```

Comment:
same here, it's super confusing to be comparing duration and id (something unique), maybe just needs better naming?